Move various operations on Words to a new `primitves` module

fjarri opened the pull request:
TODO: review function names

fjarri commented on src/primitives.rs:
```diff
@@ -0,0 +1,72 @@
+/// Computes `lhs + rhs + carry`, returning the result along with the new carry (0, 1, or 2).
+#[inline(always)]
+pub const fn adc3(lhs: Word, rhs: Word, carry: Word) -> (Word, Word) {
+    let a = lhs as WideWord;
```
This can also be expressed via `overflowing_add()`

fjarri replied:
In general, we would want to have both `adc` and `adc3` use the same method.

fjarri commented on src/primitives.rs:
```diff
@@ -0,0 +1,72 @@
+/// Computes `lhs + rhs`, returning the result along with the carry (0 or 1).
+#[inline(always)]
+pub const fn adc(lhs: Word, rhs: Word) -> (Word, Word) {
+    let (res, carry) = lhs.overflowing_add(rhs);
```
This should be constant-time according to Godbolt. Safe to use?

tarcieri replied:
The main thing that seems weird is that it doesn't accept a carry as input, so you can't actually use it to build a carry chain?

tarcieri replied:
I guess that's what you're calling `adc3` and I suppose they're both valuable

fjarri replied:
Yes, judging by the benchmarks the compiler doesn't optimize out `carry = 0`.

tarcieri replied:
Aah cool, I like this being renamed to `overflowing_add`, but in that case perhaps we can just use the method on the primitive type?

fjarri replied:
I wanted it to imply that using `Word::overflowing_add()` was "vetted" as opposed to just using it directly which may seem like an omission. Also make it easier to try different approaches e.g. via `WideWord`.

tarcieri commented:
I guess this is draft because it's WIP? Otherwise looks great so far

fjarri commented:
Adjusted the names and added some comments.

tarcieri reviewed (approved):
This seems good unless you wanted to add the Hacker's Delight overflow checks and such

fjarri commented:
If they're used elsewhere other than `const_choise.rs` we can add them, but for now I think this is good enough.

tarcieri commented:
There is duplication between `const_choice.rs` and `modular/boxed_residue/mul.rs`

fjarri commented:
Not anymore.

tarcieri commented:
It seems like these two duplicate each other:

https://github.com/RustCrypto/crypto-bigint/blob/00e1a60/src/const_choice.rs#L71-L85

And it's also duplicated here:

https://github.com/RustCrypto/crypto-bigint/blob/00e1a60/src/modular/boxed_residue/mul.rs#L305

Potentially all three could call a common function?

fjarri commented:
The one in `boxed_residue/mul` has been removed in this PR, replaced with `sbb()`. The one in `const_choice` can indeed be expressed through the existing method, fixed. The `u32`/`Word` repetitions can be removed with a macro, but I'm not sure if it's worth it.
